hax/hax/util.py

```diff
@@ -244,6 +279,28 @@ def to_payload(v: KeyDelete) -> Dict[str, Any]:
             raise HAConsistencyException(f'Failed to delete key(s)'
                                          f' from KV, error: {e}')
 
+    # Caches the result of remote KV fetch operation if not already cached.
+    # Caching is an expensive operation and is mainly required and implemented
+    # in case of expensive IO bound operations with high frequencies. It is
+    # important to understand what data must be cached and what not based on
+    # the profiling results. It is important that only high frequency IO bound
+    # operations can be cached which produce atleast 50% performance gains
+    # against their execution without caching. Invalidating the cache is also
+    # an expensive and a tricky operation. Thus candidates for caching must be
+    # carefully selected.
+    @staticmethod
+    def _kv_cache_decorator(key: str):
```

Comment:
That's a good starting point.

Suggestions:
1. Every time a function with this decorator starts, the decorator code increments a TLS counter.
2. If counter is 1, then cache must be created. The cache itself can be a TLS member of ConsulUtil object or KVAdapter.
3. Any KV reading operation should lookup firstly in the cache and if not found, then the key must be be requested from Consul. In case of actual Consul operation, the cache must be updated.
4. When such a decorated function exits, the TLS counter should be decremented
   * Note: if an exception is thrown, such decreasing must still happen. It means that decorator must also catch any exception, decrease the counter and re-throw it.
5. If the counter is 0 then the cache must be cleared (invalidated).

Reply:
I decided not to have a thread local cache. So the kvcache variable is a class variable (static) and not an instance variable. Problem with TLS is, if 1 thread updates the related consul KV, how do we invalidate the thread local caches of all the threads. That will anyway require some kind of synchronization or a signaling mechanism. I was thinking of something similar to what we have in Motr. We can add a channel corresponding to a potential event and whoever is listening on the channel will get a signal on when the event occurs. But that is more complicated.
Presently a simple static cache with locking is a simplistic approach.